File: tests/lua-tests/src/AssetsManagerExTest/AssetsManagerExTest.lua
local targetPlatform = cc.Application:getInstance():getTargetPlatform()

local sceneID = 
{
    "AMTestScene1", 
    "AMTestScene2", 
    "AMTestScene3",
}

local sceneManifests = 
{
    "Manifests/AMTestScene1/project.manifest", 
    "Manifests/AMTestScene2/project.manifest", 
    "Manifests/AMTestScene3/project.manifest"
}

local storagePaths = 
{
    "LuaTests/AssetsManagerExTest/scene1",
    "LuaTests/AssetsManagerExTest/scene2", 
    "LuaTests/AssetsManagerExTest/scene3",
}

local backgroundPaths = 
{
    "Images/assetMgrBackground1.jpg",
    "Images/assetMgrBackground2.png",
    "Images/assetMgrBackground3.png"
}

-------------------------------------
--  AssetsManagerEx1 Test
-------------------------------------
local AMTestScene1 = {}
AMTestScene1.__index = AMTestScene1

function AMTestScene1.create()

    local layer  = cc.Layer:create()

    local am = nil

    local function onEnter()

        Helper.initWithLayer(layer)
        Helper.titleLabel:setString("AssetsManagerExTest1")

        local sprite = cc.Sprite:create("Images/Icon.png")
        layer:addChild(sprite)
        sprite:setPosition(cc.p(VisibleRect:center().x, VisibleRect:center().y))

        local ttfConfig = {}
        ttfConfig.fontFilePath = "fonts/arial.ttf"
        ttfConfig.fontSize = 40

        local  progress = cc.Label:createWithTTF(ttfConfig, "0%", cc.VERTICAL_TEXT_ALIGNMENT_CENTER)
        progress:setPosition(cc.p(VisibleRect:center().x, VisibleRect:center().y + 50))
        layer:addChild(progress)
        am = cc.AssetsManagerEx:create(sceneManifests[1], cc.FileUtils:getInstance():getWritablePath() .. storagePaths[1])
        am:retain()

        if not am:getLocalManifest():isLoaded() then
            print("Fail to update assets, step skipped.")
            local background = cc.Sprite:create(backgroundPaths[1])
            layer:addChild(background, 1)
            background:setPosition( cc.p(VisibleRect:center().x, VisibleRect:center().y ))
        else
            local function onUpdateEvent(event)
                local eventCode = event:getEventCode()
                if eventCode == cc.EventAssetsManagerEx.EventCode.ERROR_NO_LOCAL_MANIFEST then
                    print("No local manifest file found, skip assets update.")
                    local background = cc.Sprite:create(backgroundPaths[1])
                    layer:addChild(background, 1)
                    background:setPosition( cc.p(VisibleRect:center().x, VisibleRect:center().y ))
                elseif  eventCode == cc.EventAssetsManagerEx.EventCode.UPDATE_PROGRESSION then
                    local assetId = event:getAssetId()
                    local percent = event:getPercent()
                    local strInfo = ""

                    if assetId == cc.AssetsManagerExStatic.VERSION_ID then
                        strInfo = string.format("Version file: %d%%", percent)
                    elseif assetId == cc.AssetsManagerExStatic.MANIFEST_ID then
                        strInfo = string.format("Manifest file: %d%%", percent)
                    else
                        strInfo = string.format("%d%%", percent)
                    end
                    progress:setString(strInfo)
                elseif eventCode == cc.EventAssetsManagerEx.EventCode.ERROR_DOWNLOAD_MANIFEST or 
                       eventCode == cc.EventAssetsManagerEx.EventCode.ERROR_PARSE_MANIFEST then
                    print("Fail to download manifest file, update skipped.")
                    local background = cc.Sprite:create(backgroundPaths[1])
                    layer:addChild(background, 1)
                    background:setPosition( cc.p(VisibleRect:center().x, VisibleRect:center().y ))
                elseif eventCode == cc.EventAssetsManagerEx.EventCode.ALREADY_UP_TO_DATE or 
                       eventCode == cc.EventAssetsManagerEx.EventCode.UPDATE_FINISHED then
                        print("Update finished.")
                        local background = cc.Sprite:create(backgroundPaths[1])
                        layer:addChild(background, 1)
                        background:setPosition( cc.p(VisibleRect:center().x, VisibleRect:center().y ))
                elseif eventCode == cc.EventAssetsManagerEx.EventCode.ERROR_UPDATING then
                        print("Asset ", event:getAssetId(), ", ", event:getMessage())
                        local background = cc.Sprite:create(backgroundPaths[1])
                        layer:addChild(background, 1)
                        background:setPosition( cc.p(VisibleRect:center().x, VisibleRect:center().y ))
                end
            end
            local listener = cc.EventListenerAssetsManagerEx:create(am,onUpdateEvent)
            cc.Director:getInstance():getEventDispatcher():addEventListenerWithFixedPriority(listener, 1)
            
            am:update()
        end
    end

    local function onNodeEvent(event)
        if "enter" == event then
            onEnter()
        elseif "exit" == event then
            am:release()
        end
    end
    layer:registerScriptHandler(onNodeEvent)

    return layer
end


-------------------------------------
--  AssetsManagerEx2 Test
-------------------------------------
local AMTestScene2 = {}
AMTestScene2.__index = AMTestScene2

function AMTestScene2.create()

    local layer  = cc.Layer:create()

    local am = nil

    local function onEnter()

        Helper.initWithLayer(layer)
        Helper.titleLabel:setString("AssetsManagerExTest2")

        local sprite = cc.Sprite:create("Images/Icon.png")
        layer:addChild(sprite)
        sprite:setPosition(cc.p(VisibleRect:center().x, VisibleRect:center().y))

        local ttfConfig = {}
        ttfConfig.fontFilePath = "fonts/arial.ttf"
        ttfConfig.fontSize = 40

        local  progress = cc.Label:createWithTTF(ttfConfig, "0%", cc.VERTICAL_TEXT_ALIGNMENT_CENTER)
        progress:setPosition(cc.p(VisibleRect:center().x, VisibleRect:center().y + 50))
        layer:addChild(progress)

        am = cc.AssetsManagerEx:create(sceneManifests[2], cc.FileUtils:getInstance():getWritablePath() .. storagePaths[2])
        am:retain()

        if not am:getLocalManifest():isLoaded() then
            print("Fail to update assets, step skipped.")
            local background = cc.Sprite:create(backgroundPaths[2])
            layer:addChild(background, 1)
            background:setPosition( cc.p(VisibleRect:center().x, VisibleRect:center().y ))
        else
            local function onUpdateEvent(event)
                local eventCode = event:getEventCode()
                if eventCode == cc.EventAssetsManagerEx.EventCode.ERROR_NO_LOCAL_MANIFEST then
                    print("No local manifest file found, skip assets update.")
                    local background = cc.Sprite:create(backgroundPaths[2])
                    layer:addChild(background, 1)
                    background:setPosition( cc.p(VisibleRect:center().x, VisibleRect:center().y ))
                elseif  eventCode == cc.EventAssetsManagerEx.EventCode.UPDATE_PROGRESSION then
                    local assetId = event:getAssetId()
                    local percent = event:getPercent()
                    local strInfo = ""

                    if assetId == cc.AssetsManagerExStatic.VERSION_ID then
                        strInfo = string.format("Version file: %d%%", percent)
                    elseif assetId == cc.AssetsManagerExStatic.MANIFEST_ID then
                        strInfo = string.format("Manifest file: %d%%", percent)
                    else
                        strInfo = string.format("%d%%", percent)
                    end
                    progress:setString(strInfo)
                elseif eventCode == cc.EventAssetsManagerEx.EventCode.ERROR_DOWNLOAD_MANIFEST or 
                       eventCode == cc.EventAssetsManagerEx.EventCode.ERROR_PARSE_MANIFEST then
                    print("Fail to download manifest file, update skipped.")
                    local background = cc.Sprite:create(backgroundPaths[2])
                    layer:addChild(background, 1)
                    background:setPosition( cc.p(VisibleRect:center().x, VisibleRect:center().y ))
                elseif eventCode == cc.EventAssetsManagerEx.EventCode.ALREADY_UP_TO_DATE or 
                       eventCode == cc.EventAssetsManagerEx.EventCode.UPDATE_FINISHED then
                        print("Update finished.")
                        local background = cc.Sprite:create(backgroundPaths[2])
                        layer:addChild(background, 1)
                        background:setPosition( cc.p(VisibleRect:center().x, VisibleRect:center().y ))
                elseif eventCode == cc.EventAssetsManagerEx.EventCode.ERROR_UPDATING then
                        print("Asset ", event:getAssetId(), ", ", event:getMessage())
                        local background = cc.Sprite:create(backgroundPaths[2])
                        layer:addChild(background, 1)
                        background:setPosition( cc.p(VisibleRect:center().x, VisibleRect:center().y ))
                end
            end
            local listener = cc.EventListenerAssetsManagerEx:create(am,onUpdateEvent)
            cc.Director:getInstance():getEventDispatcher():addEventListenerWithFixedPriority(listener, 1)
            
            am:update()
        end
    end

    local function onNodeEvent(event)
        if "enter" == event then
            onEnter()
        elseif "exit" == event then
            am:release()
        end
    end
    layer:registerScriptHandler(onNodeEvent)

    return layer
end

-------------------------------------
--  AssetsManagerEx3 Test
-------------------------------------
local AMTestScene3 = {}
AMTestScene3.__index = AMTestScene3

function AMTestScene3.create()

    local layer  = cc.Layer:create()

    local am = nil

    local function onEnter()

        Helper.initWithLayer(layer)
        Helper.titleLabel:setString("AssetsManagerExTest3")

        local sprite = cc.Sprite:create("Images/Icon.png")
        layer:addChild(sprite)
        sprite:setPosition(cc.p(VisibleRect:center().x, VisibleRect:center().y))

        local ttfConfig = {}
        ttfConfig.fontFilePath = "fonts/arial.ttf"
        ttfConfig.fontSize = 40

        local  progress = cc.Label:createWithTTF(ttfConfig, "0%", cc.VERTICAL_TEXT_ALIGNMENT_CENTER)
        progress:setPosition(cc.p(VisibleRect:center().x, VisibleRect:center().y + 50))
        layer:addChild(progress)

        am = cc.AssetsManagerEx:create(sceneManifests[3], cc.FileUtils:getInstance():getWritablePath() .. storagePaths[3])
        am:retain()

        if not am:getLocalManifest():isLoaded() then
            print("Fail to update assets, step skipped.")
            local background = cc.Sprite:create("Images/background3.png")
            layer:addChild(background, 1)
            background:setPosition( cc.p(VisibleRect:center().x, VisibleRect:center().y ))
        else
            local function onUpdateEvent(event)
                local eventCode = event:getEventCode()
                if eventCode == cc.EventAssetsManagerEx.EventCode.ERROR_NO_LOCAL_MANIFEST then
                    print("No local manifest file found, skip assets update.")
                    local background = cc.Sprite:create(backgroundPaths[3])
                    layer:addChild(background, 1)
                    background:setPosition( cc.p(VisibleRect:center().x, VisibleRect:center().y ))
                elseif  eventCode == cc.EventAssetsManagerEx.EventCode.UPDATE_PROGRESSION then
                    local assetId = event:getAssetId()
                    local percent = event:getPercent()
                    local strInfo = ""

                    if assetId == cc.AssetsManagerExStatic.VERSION_ID then
                        strInfo = string.format("Version file: %d%%", percent)
                    elseif assetId == cc.AssetsManagerExStatic.MANIFEST_ID then
                        strInfo = string.format("Manifest file: %d%%", percent)
                    else
                        strInfo = string.format("%d%%", percent)
                    end
                    progress:setString(strInfo)
                elseif eventCode == cc.EventAssetsManagerEx.EventCode.ERROR_DOWNLOAD_MANIFEST or 
                       eventCode == cc.EventAssetsManagerEx.EventCode.ERROR_PARSE_MANIFEST then
                    print("Fail to download manifest file, update skipped.")
                    local background = cc.Sprite:create(backgroundPaths[3])
                    layer:addChild(background, 1)
                    background:setPosition( cc.p(VisibleRect:center().x, VisibleRect:center().y ))
                elseif eventCode == cc.EventAssetsManagerEx.EventCode.ALREADY_UP_TO_DATE or 
                       eventCode == cc.EventAssetsManagerEx.EventCode.UPDATE_FINISHED then
                        print("Update finished.")
                        local background = cc.Sprite:create(backgroundPaths[3])
                        layer:addChild(background, 1)
                        background:setPosition( cc.p(VisibleRect:center().x, VisibleRect:center().y ))
                elseif eventCode == cc.EventAssetsManagerEx.EventCode.ERROR_UPDATING then
                        print("Asset ", event:getAssetId(), ", ", event:getMessage())
                        local background = cc.Sprite:create(backgroundPaths[3])
                        layer:addChild(background, 1)
                        background:setPosition( cc.p(VisibleRect:center().x, VisibleRect:center().y ))
                end
            end
            local listener = cc.EventListenerAssetsManagerEx:create(am,onUpdateEvent)
            cc.Director:getInstance():getEventDispatcher():addEventListenerWithFixedPriority(listener, 1)
            
            am:update()
        end
    end

    local function onNodeEvent(event)
        if "enter" == event then
            onEnter()
        elseif "exit" == event then
            am:release()
        end
    end
    layer:registerScriptHandler(onNodeEvent)

    return layer
end

-------------------------------------
--  AssetsManagerEx Test
-------------------------------------
function AssetsManagerExTestMain()
    local scene = cc.Scene:create()

    Helper.createFunctionTable = 
    {
        AMTestScene1.create,
        AMTestScene2.create,
        AMTestScene3.create,
    }

    scene:addChild(AMTestScene1.create())
    scene:addChild(CreateBackMenuItem())
    return scene
end

File: tests/cpp-tests/Classes/ExtensionsTest/AssetsManagerExTest/AssetsManagerExTest.cpp
#include "AssetsManagerExTest.h"
#include "../../testResource.h"
#include "cocos2d.h"

USING_NS_CC;
USING_NS_CC_EXT;

const char* sceneManifests[] = {"AMTestScene1/project.manifest", "AMTestScene2/project.manifest", "AMTestScene3/project.manifest"};
const char* storagePaths[] = {"CppTests/AssetsManagerExTest/scene1/", "CppTests/AssetsManagerExTest/scene2/", "CppTests/AssetsManagerExTest/scene3"};
const char* backgroundPaths[] = {"Images/assetMgrBackground1.jpg", "Images/assetMgrBackground2.png", "Images/assetMgrBackground3.png"};

AssetsManagerExTests::AssetsManagerExTests()
{
    addTestCase("AssetsManager Test1", [](){ return AssetsManagerExLoaderScene::create(0); });
    addTestCase("AssetsManager Test2", [](){ return AssetsManagerExLoaderScene::create(1); });
    addTestCase("AssetsManager Test3", [](){ return AssetsManagerExLoaderScene::create(2); });
}

AssetsManagerExLoaderScene* AssetsManagerExLoaderScene::create(int testIndex)
{
    auto scene = new (std::nothrow) AssetsManagerExLoaderScene(testIndex);
    if (scene && scene->init())
    {
        scene->autorelease();
    }
    else
    {
        delete scene;
        scene = nullptr;
    }

    return scene;
}

AssetsManagerExLoaderScene::AssetsManagerExLoaderScene(int testIndex)
: _testIndex(testIndex)
, _loadLayer(nullptr)
, _downloadMenu(nullptr)
, _am(nullptr)
, _progress(nullptr)
, _amListener(nullptr)
{

}

bool AssetsManagerExLoaderScene::init()
{
    if (!TestCase::init())
    {
        return false;
    }

    auto downloadLabel = Label::createWithTTF("Start Download", "fonts/arial.ttf", 16);
    auto downloadItem = MenuItemLabel::create(downloadLabel, CC_CALLBACK_1(AssetsManagerExLoaderScene::startDownloadCallback, this));
    downloadItem->setPosition(Vec2(VisibleRect::center().x, VisibleRect::bottom().y + 100));
    _downloadMenu = Menu::create(downloadItem, nullptr);
    _downloadMenu->setPosition(Vec2::ZERO);
    this->addChild(_downloadMenu, 1);

    _loadLayer = Layer::create();
    addChild(_loadLayer);

    auto sprite = Sprite::create("Images/Icon.png");
    _loadLayer->addChild(sprite);
    sprite->setPosition( VisibleRect::center() );
    
    TTFConfig config("fonts/tahoma.ttf", 30);
    _progress = Label::createWithTTF(config, "0%", TextHAlignment::CENTER);
    _progress->setPosition( Vec2(VisibleRect::center().x, VisibleRect::center().y + 50) );
    _loadLayer->addChild(_progress);
    
    std::string manifestPath = sceneManifests[_testIndex], storagePath = FileUtils::getInstance()->getWritablePath() + storagePaths[_testIndex];
    CCLOG("Storage path for this test : %s", storagePath.c_str());
    _am = AssetsManagerEx::create(manifestPath, storagePath);
    _am->retain();

    return true;
}

void AssetsManagerExLoaderScene::onLoadEnd()
{
    removeChild(_loadLayer, true);
    _loadLayer = nullptr;

    auto backgroundSprite = Sprite::create(backgroundPaths[_testIndex]);
    addChild(backgroundSprite, 1);
    backgroundSprite->setPosition(VisibleRect::center());
}

void AssetsManagerExLoaderScene::startDownloadCallback(Ref* sender)
{
    removeChild(_downloadMenu);
    _downloadMenu = nullptr;

    if (!_am->getLocalManifest()->isLoaded())
    {
        CCLOG("Fail to update assets, step skipped.");
        onLoadEnd();
    }
    else
    {
        int testIndex = _testIndex;
        _amListener = cocos2d::extension::EventListenerAssetsManagerEx::create(_am, [testIndex, this](EventAssetsManagerEx* event){
            static int failCount = 0;
            switch (event->getEventCode())
            {
            case EventAssetsManagerEx::EventCode::ERROR_NO_LOCAL_MANIFEST:
            {
                CCLOG("No local manifest file found, skip assets update.");
                this->onLoadEnd();
            }
            break;
            case EventAssetsManagerEx::EventCode::UPDATE_PROGRESSION:
            {
                std::string assetId = event->getAssetId();
                float percent = event->getPercent();
                std::string str;
                if (assetId == AssetsManagerEx::VERSION_ID)
                {
                    str = StringUtils::format("Version file: %.2f", percent) + "%";
                }
                else if (assetId == AssetsManagerEx::MANIFEST_ID)
                {
                    str = StringUtils::format("Manifest file: %.2f", percent) + "%";
                }
                else
                {
                    str = StringUtils::format("%.2f", percent) + "%";
                    CCLOG("%.2f Percent", percent);
                }
                if (this->_progress != nullptr)
                    this->_progress->setString(str);
            }
            break;
            case EventAssetsManagerEx::EventCode::ERROR_DOWNLOAD_MANIFEST:
            case EventAssetsManagerEx::EventCode::ERROR_PARSE_MANIFEST:
            {
                CCLOG("Fail to download manifest file, update skipped.");
                this->onLoadEnd();
            }
            break;
            case EventAssetsManagerEx::EventCode::ALREADY_UP_TO_DATE:
            case EventAssetsManagerEx::EventCode::UPDATE_FINISHED:
            {
                CCLOG("Update finished. %s", event->getMessage().c_str());
                this->onLoadEnd();
            }
            break;
            case EventAssetsManagerEx::EventCode::UPDATE_FAILED:
            {
                CCLOG("Update failed. %s", event->getMessage().c_str());

                failCount++;
                if (failCount < 5)
                {
                    _am->downloadFailedAssets();
                }
                else
                {
                    CCLOG("Reach maximum fail count, exit update process");
                    failCount = 0;
                    this->onLoadEnd();
                }
            }
            break;
            case EventAssetsManagerEx::EventCode::ERROR_UPDATING:
            {
                CCLOG("Asset %s : %s", event->getAssetId().c_str(), event->getMessage().c_str());
            }
            break;
            case EventAssetsManagerEx::EventCode::ERROR_DECOMPRESS:
            {
                CCLOG("%s", event->getMessage().c_str());
            }
            break;
            default:
                break;
            }
        });
        Director::getInstance()->getEventDispatcher()->addEventListenerWithFixedPriority(_amListener, 1);

        _am->update();
    }
}

void AssetsManagerExLoaderScene::onExit()
{
    _eventDispatcher->removeEventListener(_amListener);
    _am->release();
    TestCase::onExit();
}

std::string AssetsManagerExLoaderScene::title() const
{
    return "AssetsManagerExTest";
}
